movement-test/src/main.c

```diff
@@ -64,80 +78,306 @@ UINT8 update_sprite_animation(UINT8 sprite_id, UINT8 *anim, UINT8 direction, UIN
+      dx = dx + dx;
+      dy = dy + dy;
+    }
+    if (compteur_dash == step3_dash){
+      // The caracter slows down at the end of the dash
+
+      if (dx > 30) dx = dx + 2;
+      if (dx > 0 && dx < 30) dx = dx -2;
+      if (dy > 30) dy = dy + 2;
+      if (dy > 0 && dx < 30) dy = dy -2;
+    }
+
+    //move_sprite(PLAYER_SPRITE_ID, player.pos.x, player.pos.y);
+    compteur_dash += 1;
+  }
+}
+
+
+void handle_collisions() {
+  total_diff.dx = dx;
+  total_diff.dy = dy;
+
+  while (total_diff.dx != 0 || total_diff.dy != 0) {
+    // Split movement into submoves (to avoid going past obstacles)
+
+    // XXX: the generated assembly for this lines is ridiculously
+    // long IMO (~104 instructions per line!).
+    current_diff.dx = ABS_CLAMP(total_diff.dx, MAX_MOVE_X);
+
+    total_diff.dx -= current_diff.dx;
+    new_player.pos.x = player.pos.x + current_diff.dx;
+
+
+    // Player's position using tilemap coordinates
+    effective_x = (new_player.pos.x >> 3);
+    effective_y = (player.pos.y >> 3);
+
+    for (UINT8 i = 0; i < 3; i++) {
+      for (UINT8 j = 0; j < 3; j++) {
+        INT8 block_x = (effective_x + i - 1);
+        INT8 block_y = (effective_y + j - 1);
+
+        if (block_x < 0 || block_x >= (INT8) ROOM_WIDTH + 2 || block_y < 0 || block_y >= (INT8) ROOM_HEIGHT + 2)  // out of bounds
+          continue;
+
+        INT16 k = block_y * (ROOM_WIDTH + 2) + block_x;
+
+        if (TILEMAP[k] != 0) {
+          block.pos.x = block_x << 3;
+          block.pos.y = block_y << 3;
+
+          if (rect_rect_collision(&new_player, &block)) {
+            VEC_DIFF diff = {0, 0};
+            rect_rect_penetration(&(player.pos), &(new_player.pos), &(player.size), &block, &diff);
+            new_player.pos.x += diff.dx;
+            total_diff.dx = 0;
+          }
+        }
+      }
+    }
+
+    current_diff.dy = ABS_CLAMP(total_diff.dy, MAX_MOVE_Y);
+    total_diff.dy -= current_diff.dy;
+    new_player.pos.y = player.pos.y + current_diff.dy;
+
+    effective_x = (new_player.pos.x >> 3);
+    effective_y = (new_player.pos.y >> 3);
+
+    for (UINT8 i = 0; i < 3; i++) {
+      for (UINT8 j = 0; j < 3; j++) {
+        INT8 block_x = (effective_x + i - 1);
+        INT8 block_y = (effective_y + j - 1);
+
+        if (block_x < 0 || block_x >= (INT8) ROOM_WIDTH + 2 || block_y < 0 || block_y >= (INT8) ROOM_HEIGHT + 2)  // out of bounds
+          continue;
+
+        INT16 k = block_y * (ROOM_WIDTH + 2) + block_x;
+
+        if (TILEMAP[k] != 0) {
+          block.pos.x = block_x << 3;
+          block.pos.y = block_y << 3;
+
+          if (rect_rect_collision(&new_player, &block)) {
+            VEC_DIFF diff = {0, 0};
+            rect_rect_penetration(&(player.pos), &(new_player.pos), &(player.size), &block, &diff);
+            new_player.pos.y += diff.dy;
+            total_diff.dy = 0;
+          }
+        }
+      }
+    }
+
+    player.pos.x = new_player.pos.x;
+    player.pos.y = new_player.pos.y;
```

Comment:
We could avoid duplication here with a macro

Reply:
If you know how to write it, I would gladly accept it!

Reply:
I'll try to have a go at it